File: csharp/src/MultiAgentCopilot/Common/Models/Chat/Session.cs
using Newtonsoft.Json;

namespace MultiAgentCopilot.Common.Models.Chat;

public record Session
{
    /// <summary>
    /// Unique identifier
    /// </summary>
    public string Id { get; set; }

    public string Type { get; set; }

    /// <summary>
    /// Partition key
    /// </summary>
    public string SessionId { get; set; }

    public string TenantId { get; set; }

    public string UserId { get; set; }

    public string Name { get; set; }

    [JsonIgnore]
    public List<Message> Messages { get; set; }

    public Session(string tenantId, string userId)
    {
        Id = Guid.NewGuid().ToString();
        TenantId = tenantId;
        UserId = userId;
        Type = nameof(Session);
        SessionId = Id;
        Name = "New Chat";
        Messages = new List<Message>();
    }

    public void AddMessage(Message message)
    {
        Messages.Add(message);
    }

    public void UpdateMessage(Message message)
    {
        var match = Messages.Single(m => m.Id == message.Id);
        var index = Messages.IndexOf(match);
        Messages[index] = message;
    }
}

File: 02_completed/frontend/src/app/models/session.ts
export class Session {
    id: string;
    type: string;
    tenantId: string;
    userId: string;
    sessionId: string;
    tokens: number | null;
    name: string;
    private messages: Message[];

    constructor(tenantId: string, userId: string, name: string, sessionId?: string) {
        this.id = this.generateUUID();
        this.type = "Session";
        this.sessionId = sessionId ?? this.generateUUID();
        this.userId = userId;
        this.tenantId = tenantId;
        this.tokens = 0;
        this.name = name
        this.messages = [];
    }

    private generateUUID(): string {
        return crypto.randomUUID();
    }

    addMessage(message: Message): void {
        this.messages.push(message);
    }

    updateMessage(message: Message): void {
        const index = this.messages.findIndex(m => m.id === message.id);
        if (index !== -1) {
            this.messages[index] = message;
        }
    }
}

interface Message {
    id: string;
    content: string;
}
